src/libs/elektra/keyhelpers.c

```diff
@@ -197,10 +197,13 @@ int keyNameIsSystem (const char * name)
  */
 int keyInit (Key * key)
 {
-	if (key->meta) ksClear (key->meta);
-	memset (key, 0, sizeof (struct _Key) - sizeof (KeySet *));
+	KeySet * ks = key->meta;
+	if (ks) ksClear (ks);
 
-	if (!key->meta) key->meta = ksNew (0, KS_END);
+	memset (key, 0, sizeof (struct _Key));
+	key->meta = ks;
```

Comment:
I had a simpler solution before, that did not require the `NULL` precondition for `key->meta`. It works fine for the abi, but the `mmapstorage` plugin test will not work with that change and I could not find the source of the error in reasonable time, so I rolled some of those changes back.

Reply:
If there's a problem with `mmapstorage` please report it, maybe I can help or take a look.

I don't get why `keyInit` requires changes here.

I think that use of an [uninitialized stack variable](https://cwe.mitre.org/data/definitions/457.html) is introduced here.  `ksLookupByName` makes a call to `keyInit` by supplying an uninitialized Key struct on the stack. To my knowledge this means that some garbage from memory will be passed to `keyInit` and lead to calling `ksClear` on an invalid pointer.

Reply:
This is a continuation of #3081 where I already made changes to `keyInit` to initialize the `key->meta` on key creation. But the implementation wasn't portable and not ideal so I changed it again in this PR.

In #3081 I changed `ksLookupByName` to initialize the meta variable to null, so that shouldn't be an issue.

If you could take a look that would be great. The breaking change was introduced in 
1130515. Maybe check that out and run `test_mmapstorage`. You can also see it [here on Travis](https://travis-ci.org/ElektraInitiative/libelektra/jobs/605429496?utm_medium=notification&utm_source=github_status).

Reply:
Thank you, I will take a look!

Reply:
I took a look at it and the problem with `mmapstorage` is caused by the allocation of the empty meta keyset. 

I attached a patch with changes that make 11305155ac3efc58357741161d827c60af280b4a work with mmapstorage, lazily create the meta keyset and use elektraCalloc as suggested in my previous comments. (The .patch is renamed to .txt because github doesnt like .patch files)
[keyMeta-lazy.txt](https://github.com/ElektraInitiative/libelektra/files/3800466/keyMeta-lazy.txt)

Reply:
Thank you, works great!